ignite/contrib/handlers/lr_finder.py

```diff
@@ -0,0 +1,405 @@
+
+        # attach loss and lr logging
+        if not engine.has_event_handler(self._log_lr_and_loss):
+            engine.add_event_handler(Events.ITERATION_COMPLETED, self._log_lr_and_loss, smooth_f, diverge_th)
+
+        self._logger.debug("Running LR finder for {} iterations".format(num_iter))
+        # Initialize the proper learning rate policy
+        if step_mode.lower() == "exp":
+            self._lr_schedule = LRScheduler(_ExponentialLR(self._optimizer, end_lr, num_iter))
+        else:
+            start_lr = self._optimizer.param_groups[0]["lr"]
+            self._lr_schedule = PiecewiseLinear(self._optimizer, "lr", [(0, start_lr), (num_iter, end_lr)])
+        if not engine.has_event_handler(self._lr_schedule):
+            engine.add_event_handler(Events.ITERATION_COMPLETED, self._lr_schedule, num_iter)
+
+    # Reset model and optimizer, delete state cache and remove handlers
+    def _reset(self, engine):
+        self._model.load_state_dict(engine.state.state_cache.retrieve("model"))
+        self._optimizer.load_state_dict(engine.state.state_cache.retrieve("optimizer"))
+        del engine.state.state_cache
+        self._logger.debug("Completed LR finder run, resets model & optimizer")
+
+        # Clean up; remove event handlers added during run
+        engine.remove_event_handler(self._lr_schedule, Events.ITERATION_COMPLETED)
+        engine.remove_event_handler(self._log_lr_and_loss, Events.ITERATION_COMPLETED)
+        engine.remove_event_handler(self._reached_num_iterations, Events.ITERATION_COMPLETED)
+
+    def _log_lr_and_loss(self, engine, smooth_f, diverge_th):
+        output = engine.state.output
+        loss = self._output_transform(output)
+        lr = self._lr_schedule.get_param()
+        self._history["lr"].append(lr)
+        if engine.state.iteration == 1:
+            self._best_loss = loss
+        else:
+            if smooth_f > 0:
+                loss = smooth_f * loss + (1 - smooth_f) * self._history["loss"][-1]
+            if loss < self._best_loss:
+                self._best_loss = loss
+        self._history["loss"].append(loss)
+
+        # Check if the loss has diverged; if it has, stop the trainer
+        if self._history["loss"][-1] > diverge_th * self._best_loss:
+            self._diverge_flag = True
+            self._logger.info("Stopping early, the loss has diverged")
+            engine.terminate()
+
+    def _reached_num_iterations(self, engine, num_iter):
+        if engine.state.iteration > num_iter:
+            engine.terminate()
+
+    def _warning(self, engine):
+        if not self._diverge_flag:
+            warnings.warn("Run completed without loss diverging, increase end_lr, decrease diverge_th or look"
+                          " at lr_finder.plot()", UserWarning)
+
+    def attach(self, engine, model, optimizer, output_transform=lambda output: output, num_iter=None, end_lr=10,
```

Comment:
Shouldn't it be possible to achieve the context manager behaviour with the `context manager` decorator from [contextlib](https://docs.python.org/3/library/contextlib.html) like the following:

```python
from context lib import contextmanager

@contextmanager
def attach(self, engine, model, optimizer, output_transform=lambda output: output, num_iter=None, end_lr=10, step_mode="exp", smooth_f=0.05, diverge_th=5):
# DO ALL THE ATTACHING STUFF
    yield self
    self.detach(self._engine)

Reply:
@justusschock could you please test if we could achieve that with contextlib ?
Yes, we can yield engine and detail it as :
```python
with lr_finder.attach(trainer) as trainer_to_find_lr:
    trainer_to_find_lr.run(dataloader)
```

Reply:
Oh, didn't know about `contextlib` until now... Thanks @justusschock !!!
I tried it now and it seems to work. I can finish it and fix the tests and documentation if you want

Reply:
@ItamarWilf yes, please go ahead